src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt

```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
             sv.analyzer = aa
             aa
         }
+        mv = CoverageBytecodeFilter(
+            skipVisitor.newAdapter(),
+            mv.newAdapter()
+        )
         mv = DeterministicHashCodeTransformer(fileName, className, methodName, mv.newAdapter())
         mv = DeterministicTimeTransformer(mv.newAdapter())
         mv = DeterministicRandomTransformer(fileName, className, methodName, mv.newAdapter())
         return mv
     }
 
 }
+    /**
+     * Filters out chunks of bytecode generated by coverage.
+     *
+     * @param initialAdapter initial adapter to delegate work to when coverage-related bytecode encountered.
+     * @param nextAdapter adapter to delegate calls to when bytecode is not related to coverage.
+     */
+    private class CoverageBytecodeFilter(
+        private val initialAdapter: GeneratorAdapter,
+        nextAdapter: GeneratorAdapter
+    ) : MethodVisitor(ASM_API, nextAdapter) {
+        enum class State {
+            INITIAL,
+            LOADED_HITS_ON_STACK,
+            LOADED_HITS_ON_STACK_STATIC,
+            LOADED_HITS_IN_LOCALS,
+            PREPARE_ASSIGN_TO_HITS,
+        }
+        private val COVERAGE_HITS_NAME = "__\$hits\$__"
+        private val COVERAGE_GET_HITS_MASK_METHOD_SIGNATURE = "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMask(Ljava/lang/String;)[Z"
```

Comment:
For the `hits[i]++` version, here will be 
```
com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHits (Ljava/lang/String;)[I
```
You can check all options here https://github.com/JetBrains/intellij-coverage/tree/master/test-kotlin/resources/bytecode/simple/branches